xds/internal/xdsclient/e2e_test/cds_watchers_test.go

```diff
@@ -0,0 +1,869 @@
+	})
+	defer cdsCancel1()
+	resourceName2 := cdsNameNewStyle
+	updateCh2 := testutils.NewChannel()
+	cdsCancel2 := client.WatchCluster(resourceName2, func(u xdsresource.ClusterUpdate, err error) {
+		updateCh2.Send(xdsresource.ClusterUpdateErrTuple{Update: u, Err: err})
+	})
+	defer cdsCancel2()
+
+	// Configure the management server to return two cluster resources,
+	// corresponding to the registered watches.
+	resources := e2e.UpdateOptions{
+		NodeID: nodeID,
+		Clusters: []*v3clusterpb.Cluster{
+			e2e.DefaultCluster(resourceName1, edsName, e2e.SecurityLevelNone),
+			e2e.DefaultCluster(resourceName2, edsNameNewStyle, e2e.SecurityLevelNone),
+		},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Verify the contents of the received update for both watchers.
+	wantUpdate1 := xdsresource.ClusterUpdateErrTuple{
+		Update: xdsresource.ClusterUpdate{
+			ClusterName:    resourceName1,
+			EDSServiceName: edsName,
+		},
+	}
+	wantUpdate2 := xdsresource.ClusterUpdateErrTuple{
+		Update: xdsresource.ClusterUpdate{
+			ClusterName:    resourceName2,
+			EDSServiceName: edsNameNewStyle,
+		},
+	}
+	if err := verifyClusterUpdate(ctx, updateCh1, wantUpdate1); err != nil {
+		t.Fatal(err)
+	}
+	if err := verifyClusterUpdate(ctx, updateCh2, wantUpdate2); err != nil {
+		t.Fatal(err)
+	}
+
+	// Remove the first cluster resource on the management server.
+	resources = e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{e2e.DefaultCluster(resourceName2, edsNameNewStyle, e2e.SecurityLevelNone)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// The first watcher should receive a resource removed error, while the
+	// second watcher should not see an update.
+	if err := verifyClusterUpdate(ctx, updateCh1, xdsresource.ClusterUpdateErrTuple{Err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "")}); err != nil {
+		t.Fatal(err)
+	}
+	if err := verifyNoClusterUpdate(ctx, updateCh2); err != nil {
+		t.Fatal(err)
+	}
+
+	// Update the second cluster resource on the management server. The first
+	// watcher should not see an update, while the second watcher should.
+	resources = e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{e2e.DefaultCluster(resourceName2, "new-eds-resource", e2e.SecurityLevelNone)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+	if err := verifyNoClusterUpdate(ctx, updateCh1); err != nil {
```

Comment:
Same comment as previously, please switch the returned error to explicitly state expecting no cluster update. This keeps it consistent to the got/want/diffs we all know and love.

Reply:
As stated in the previous comment about the same thing, `unexpected update` means the same as `unexpected update when we don't want one`. We use this pattern quite a lot in our tests. Like we don't have a `want` section here. We have received something when we dont expect to receive anything. So, I feel this is good enough. Let me know if you feel strongly about it.

Reply:
Commented on other one.